tests/ui/cypress/specs/multi_user_openid_connect.js

```diff
@@ -0,0 +1,289 @@
+    cy.contains('Mock OpenID Connect login form')
+
+    // check that a username input field is shown on the page
+    cy.get('input[name="username"]')
+  })
+
+  login_test_settings.forEach(function (ts) {
+    it(
+      'Login with ' +
+        ts.d +
+        ' credentials should ' +
+        (ts.o ? 'succeed with the expected user info' : 'fail with the expected error'),
+      () => {
+        cy.visit('/')
+        cy.url().should('not.include', Cypress.config('baseUrl'))
+        cy.contains('Mock OpenID Connect login form')
+
+        if (ts.u != '') cy.get('input[name="username"]').clear().type(ts.u)
+
+        cy.contains('Sign In').click()
+
+        // We should end up back in the Krill UI
+        cy.url().should('include', Cypress.config('baseUrl'))
+
+        if (ts.o) {
+          cy.contains('Sign In').should('not.exist')
+          cy.get('#userinfo').click()
+          cy.get('#userinfo_table').contains(ts.u)
+          cy.get('#userinfo_table').contains('role')
+        } else if (ts.d == 'empty') {
+          cy.contains('The supplied login credentials were incorrect')
+          cy.contains('return to the login page')
+        } else if (ts.d == 'badidtoken') {
+          cy.contains('OpenID Connect: Code exchange failed: Failed to parse server response')
+          cy.contains('return to the login page')
+        } else if (ts.d == 'badrole') {
+          cy.contains(
+            'Your user does not have sufficient rights to perform this action. Please contact your administrator.'
+          )
+          cy.contains('return to the login page')
+        }
+      }
+    )
+  })
+
+  it('Can logout', () => {
+    // login
+    cy.visit('/')
+    cy.url().should('not.include', Cypress.config('baseUrl'))
+    cy.contains('Mock OpenID Connect login form')
+    cy.get('input[name="username"]').clear().type(admin.u)
+    cy.contains('Sign In').click()
+
+    // verify that we are shown to be logged in to the Krill UI
+    cy.contains('Sign In').should('not.exist')
+    cy.url().should('include', Cypress.config('baseUrl'))
+    cy.get('#userinfo').click()
+    cy.get('#userinfo_table').contains(admin.u)
+
+    // logout
+    cy.intercept('GET', /^http:\/\/localhost:1818\/logout.+/).as('oidcLogout')
+    cy.get('.logout').click()
+    cy.wait('@oidcLogout').its('response.statusCode').should('eq', 302)
+
+    // verify that we are shown the OpenID Connect provider login page
+    cy.url().should('not.include', Cypress.config('baseUrl'))
+    cy.contains('Mock OpenID Connect login form')
+    cy.get('input[name="username"]')
+  })
+
+  it('Login with short-lived non-refreshable token and try to refresh page', () => {
+    cy.intercept('GET', '/api/v1/authorized').as('isAuthorized')
+    cy.visit('/')
+
+    cy.wait('@isAuthorized').its('response.statusCode').should('eq', 403)
+    cy.url().should('not.include', Cypress.config('baseUrl'))
+    cy.contains('Mock OpenID Connect login form')
```

Comment:
You added 'try to create a CA" to the name of the next test but also extended this test to try to create a CA. Actually, I must be missing something, how does this test differ from the next one? What is meant by `[valid refresh]` ? I don't see any difference in the logic of this test and the next except that this test also checks that the user is shown the OIDC login page at the end of the test?

Reply:
You're right, I copied and extended the first one, but then left the original one there. There's a slight difference though, the first one (starting 126 'Login with short-lived non-refreshable token and try to refresh page') tries to refresh the page, a legit action that will receive a redirect back to login, whereas the second one (starting 179 '\[valid refresh\] Login with short-lived non-refreshable token and try to create a CA') does not refresh and therefore receives an API error and will not redirect to login.

The \[valid refresh\] is wrong though, I'll remove it, should be \[invalid refresh\].